File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    map<int, int> rightView;
    vector<int> answer;
    void dfs(TreeNode* root, int level){
        if(root == NULL)
            return;
        if(rightView.find(level) == rightView.end()){
            rightView[level] = 1;
            answer.push_back(root->val);
        }
        dfs(root->right, level + 1);
        dfs(root->left, level + 1);
    }
    vector<int> rightSideView(TreeNode* root) {
        dfs(root, 0);
        return answer;
    }
};

File: 0199-binary-tree-right-side-view/0199-binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# class Solution:
#     def __init__(self):
#         self.res = []
        
#     def dfs(self, root, level):
#         if root is None:
#             return
#         if level == len(self.res):
#             self.res.append(root.val)
#         self.dfs(root.right, level + 1)
#         self.dfs(root.left, level + 1)
        
#     def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
#         self.dfs(root, 0)
#         return self.res
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        res = []
        queue = deque([root])
        while queue:
            qlen = len(queue)
            right_view = None
            for i in range(qlen):
                node = queue.popleft()
                if node:
                    right_view = node
                    queue.append(node.left)
                    queue.append(node.right)
            if right_view:
                res.append(right_view.val)
        return res
                